### ISSUE TITLE
Compilation failure on OpenBSD (crash_handler_x11.cpp:56: undefined reference to `backtrace_symbols')

### ISSUE THREAD

**Issue by AbandonedCode** [godotengine/godot#11581]
I'd like to preface this by saying YES, godot is compatible with OpenBSD. I've successfully compiled 2.1.3 in the past.

-----

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
OpenBSD Current AMD64 from 2017/09/24 
Scons V2.5.1
LLVM 5.0.0
libexecinfo 0.3
Latest Godot from Github
 
**Issue description:**

By default, OpenBSD stores system libraries in _/usr/include_ and all other libraries in _/usr/local/include_. By default, Clang doesn't look for libraries in _/usr/local/include_ but I've added _C_INCLUDE_PATH_ in my _~/.profile_ to account for this. However, it seems Scons isn't passing these environment variable to Clang which fails to find the libraries for compilation.

To demonstrate this behaviour, If I was to write a small test program

> #include "execinfo.h"
> #include "stdio.h"
> int main() {
> printf("Hello World\n");
> return(0); }

and compile it without the environment variable added in my .profile, Clang would fail to find any libraries that are not part of OpenBSD's base system libraries.

> $ clang test.c 
> test.c:1:10: fatal error: 'execinfo.h' file not found
> #include "execinfo.h"
>          ^~~~~~~~~~~~
> 1 error generated.

After I added the environment variable in my .profile, Clang compiles without issue. 

> $ clang test.c
> $ file a.out
> a.out: ELF 64-bit LSB shared object, x86-64, version 1

Despite the environment variable being in my profile and being passed through Scons, compilation fails due to Clang not finding the libraries in question.

> $ file /usr/local/include/execinfo.h                                            
> /usr/local/include/execinfo.h: ASCII C program text
> $ scons platform=x11 use_llvm=yes CCFLAGS="/usr/local/include" CFLAGS="/usr/local/include" C_INCLUDE_PATH="/usr/local/include"
> scons: Reading SConscript files ...
> Could not detect MinGW. Ensure its binaries are in your PATH or that MINGW32_PREFIX or MINGW64_PREFIX are properly defined.
> ALSA libraries not found, disabling driver
> PulseAudio development libraries not found, disabling driver
> cdir is: /home/gjp/Git/godot
> Checking for C header file mntent.h... no
> scons: done reading SConscript files.
> scons: Building targets ...
> [Initial build] Compiling ==> platform/x11/godot_x11.cpp
> [Initial build] Compiling ==> platform/x11/context_gl_x11.cpp
> [Initial build] Compiling ==> platform/x11/crash_handler_x11.cpp
> platform/x11/crash_handler_x11.cpp:40:10: fatal error: 'execinfo.h' file not found
> #include <execinfo.h>
>          ^~~~~~~~~~~~
> 1 error generated.
> scons: *** [platform/x11/crash_handler_x11.x11.tools.64.llvm.o] Error 1
> scons: building terminated because of errors.
> 

**Steps to reproduce:**
Install [OpenBSD Current](https://ftp.openbsd.org/pub/OpenBSD/snapshots/amd64/) on a VM or bare metal. Remember to install ALL package sets. If you have any trouble, [consult the installation guide](https://www.openbsd.org/faq/faq4.html).

Great! Now as root, verify the installpath and install libexecinfo, llvm, scons and git:

`cat /etc/installurl`
`https://ftp.openbsd.org/pub/OpenBSD`
`pkg_add git llvm libexecinfo scons`

Then, clone Godot git repo

`git clone https://github.com/godotengine/godot
`
Don't forget to add C_INCLUDE_PATH to your .profile, you should also relogin for changes to take effect. 

`echo "export C_INCLUDE_PATH=/usr/local/include" >> ~/.profile
`
Finally, start the compilation

`scons platform=x11 use_llvm=yes builtin_openssl=yes`

**Comment by akien-mga** [comment]
> scons platform=x11 use_llvm=yes CCFLAGS="/usr/local/include"

Your CCFLAGS are incorrect, it should be `CCFLAGS="-I/usr/local/include"` to add a custom include path.

**Comment by AbandonedCode** [comment]
I made a change to SConstruct which (I know breaks cross-compatibility) allowed the compiler to find the libraries. I'm currently having trouble getting clang's linker to find the libraries.

```
[Initial build] Linking Program        ==> bin/godot.x11.tools.64.llvm
/usr/X11R6/lib/libXcursor.so.5.0: warning: warning: sprintf() is often misused, please use snprintf()
modules/libmodules.x11.tools.64.llvm.a(ssl_lib.x11.tools.64.llvm.o): In function `SSL_get_shared_ciphers':
thirdparty/openssl/ssl/ssl_lib.c:1434: warning: warning: strcpy() is almost always misused, please use strlcpy()
/usr/X11R6/lib/libGL.so.17.1: warning: warning: random() may return deterministic values, is that what you want?
modules/libmodules.x11.tools.64.llvm.a(v3_alt.x11.tools.64.llvm.o): In function `i2v_GENERAL_NAME':
thirdparty/openssl/crypto/x509v3/v3_alt.c:167: warning: warning: strcat() is almost always misused, please use strlcat()
core/libcore.x11.tools.64.llvm.a(zip.x11.tools.64.llvm.o): In function `crypthead':
thirdparty/minizip/crypt.h:117: warning: warning: rand() may return deterministic values, is that what you want?
platform/x11/crash_handler_x11.x11.tools.64.llvm.o: In function `_ZL12handle_crashi':
platform/x11/crash_handler_x11.cpp:49: undefined reference to `backtrace'
platform/x11/crash_handler_x11.cpp:56: undefined reference to `backtrace_symbols'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.x11.tools.64.llvm] Error 1
scons: building terminated because of errors.
```

apply patch with `patch -p0 < openbsd.patch`
```
--- SConstruct  Tue Sep 26 09:59:08 2017
+++ SConstruct  Tue Sep 26 09:59:59 2017
@@ -70,6 +70,8 @@ if 'TERM' in os.environ:
     env_base['ENV']['TERM'] = os.environ['TERM']
 env_base.AppendENVPath('PATH', os.getenv('PATH'))
 env_base.AppendENVPath('PKG_CONFIG_PATH', os.getenv('PKG_CONFIG_PATH'))
+env_base.Append( CPPFLAGS=['-I/usr/local/include'] )
+env_base.Append( LDFLAGS=['-L/usr/local/include'] )
 env_base.global_defaults = global_defaults
 env_base.android_maven_repos = []
 env_base.android_dependencies = []
```

**Comment by akien-mga** [comment]
> `+env_base.Append( LDFLAGS=['-L/usr/local/include'] )`

I doubt OpenBSD packs its libraries in `/usr/local/include` - this folder should only contain headers (.h and .hpp). The libraries (.so) should be somewhere else, probably `/usr/local/lib`.

Instead of patching SConstruct, just do `scons CCFLAGS="-I/usr/local/include" LINKFLAGS="-L/usr/local/lib"`.

**Comment by AbandonedCode** [comment]
I recompiled with an unpatched SConstruct with the arguments pointing properly towards the libraries (hehe, I was lacking morning coffee). Unfortunately, clang's linker is still having the same trouble.

```
$ scons platform=x11 use_llvm=yes builtin_openssl=yes CCFLAGS="-I/usr/local/include" LINKFLAGS="-L/usr/local/lib" 
scons: Reading SConscript files ...
Could not detect MinGW. Ensure its binaries are in your PATH or that MINGW32_PREFIX or MINGW64_PREFIX are properly defined.
ALSA libraries not found, disabling driver
PulseAudio development libraries not found, disabling driver
cdir is: /home/gjp/Git/godot
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[Initial build] Linking Program        ==> bin/godot.x11.tools.64.llvm
/usr/X11R6/lib/libXcursor.so.5.0: warning: warning: sprintf() is often misused, please use snprintf()
modules/libmodules.x11.tools.64.llvm.a(ssl_lib.x11.tools.64.llvm.o): In function `SSL_get_shared_ciphers':
thirdparty/openssl/ssl/ssl_lib.c:1434: warning: warning: strcpy() is almost always misused, please use strlcpy()
/usr/X11R6/lib/libGL.so.17.1: warning: warning: random() may return deterministic values, is that what you want?
modules/libmodules.x11.tools.64.llvm.a(v3_alt.x11.tools.64.llvm.o): In function `i2v_GENERAL_NAME':
thirdparty/openssl/crypto/x509v3/v3_alt.c:167: warning: warning: strcat() is almost always misused, please use strlcat()
core/libcore.x11.tools.64.llvm.a(zip.x11.tools.64.llvm.o): In function `crypthead':
thirdparty/minizip/crypt.h:117: warning: warning: rand() may return deterministic values, is that what you want?
platform/x11/crash_handler_x11.x11.tools.64.llvm.o: In function `_ZL12handle_crashi':
platform/x11/crash_handler_x11.cpp:49: undefined reference to `backtrace'
platform/x11/crash_handler_x11.cpp:56: undefined reference to `backtrace_symbols'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.x11.tools.64.llvm] Error 1
scons: building terminated because of errors.
```

**Comment by akien-mga** [comment]
CC @marcelofg55. Maybe the `crash_handle_x11.cpp` code was written with Linux in mind and needs some fixes for *BSD.

**Comment by AbandonedCode** [comment]
@akien-mga I'm not sure it's `crash_handle_x11.cpp` at fault, backtrace and backtrace_symbols are present in both headers and compiled libraries (verified with strings). For more clarity into what might be going on, I've ran Scons with verbose=1

`[Initial build] clang++ -o bin/godot.x11.tools.64.llvm -rdynamic -pipe -rdynamic platform/x11/godot_x11.x11.tools.64.llvm.o platform/x11/context_gl_x11.x11.tools.64.llvm.o platform/x11/crash_handler_x11.x11.tools.64.llvm.o platform/x11/os_x11.x11.tools.64.llvm.o platform/x11/key_mapping_x11.x11.tools.64.llvm.o platform/x11/joypad_linux.x11.tools.64.llvm.o platform/x11/power_x11.x11.tools.64.llvm.o -L/usr/X11R6/lib -L/usr/local/lib main/libmain.x11.tools.64.llvm.a main/tests/libtests.x11.tools.64.llvm.a modules/libmodules.x11.tools.64.llvm.a drivers/libdrivers.x11.tools.64.llvm.a editor/libeditor.x11.tools.64.llvm.a scene/libscene.x11.tools.64.llvm.a servers/libservers.x11.tools.64.llvm.a core/libcore.x11.tools.64.llvm.a modules/svg/libsvg_builtin.x11.tools.64.llvm.a -lXcursor -lXfixes -lXinerama -lXrandr -lXext -lXrender -lX11 -lxcb -lpthread-stubs -lXau -lXdmcp -lfreetype -lpng -lz -lGL -lpthread modules/recast/librecast_builtin.x11.tools.64.llvm.a`

**Comment by marcelofg55** [comment]
I'm guessing it misses a -lexecinfo or -lbacktrace perhaps?

**Comment by akien-mga** [comment]
Yeah this [bug report](http://trac.wxwidgets.org/ticket/9783) hints that `-lexecinfo` might be needed on FreeBSD - maybe similar on OpenBSD.

**Comment by AbandonedCode** [comment]
I've managed to compile commit f577efd47ed2f0aa801155756e4ba5d5ea2d5f2b on OpenBSD with the following patch. I haven't been able to test for breakage on other platforms yet. Strangely, after applying this patch Scons has now started to report a percentage during compilation.

Compiled after applying the patch using:

>  $ scons platform=x11 use_llvm=yes builtin_openssl=yes CCFLAGS="-I/usr/local/include"

```
--- platform/x11/SCsub	Tue Sep 26 16:48:48 2017
+++ platform/x11/SCsub	Tue Sep 26 18:10:08 2017
@@ -17,6 +17,8 @@ common_x11 = [
     "power_x11.cpp",
 ]
 
+env.Append(LINKFLAGS=['-lexecinfo'])
+
 binary = env.Program('#bin/godot', ['godot_x11.cpp'] + common_x11)
 if env["debug_symbols"] == "full" or env["debug_symbols"] == "yes":
     env.AddPostAction(binary, make_debug)
```

**Comment by akien-mga** [comment]
This was fixed a while ago by 2baf54526b.
